### ISSUE TITLE
Lots of "Reserved bits must be 0" messages

### ISSUE THREAD

**Issue by jacksonofalltrades** [aaugustin/websockets#290]
Hi,

I have a staging system with a websocket server set up and I am getting tons of notifications of exceptions like this:

```
asyncio base_events: Task exception was never retrieved
future: <Task finished coro=<WebSocketCommonProtocol.run() done, defined at /usr/local/lib/python3.5/dist-packages/websockets/protocol.py:428> exception=InvalidState('Cannot close a WebSocket connection in the CONNECTING state',) created at /usr/local/lib/python3.5/dist-packages/websockets/protocol.py:660>
source_traceback: Object created at (most recent call last):
  File "/srv/locix/locix/websocket/server.py", line 326, in <module>
    main(args.listen, args.port)
  File "/srv/locix/locix/websocket/server.py", line 307, in main
    main_loop.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1304, in _run_once
    handle._run()
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/local/lib/python3.5/dist-packages/websockets/server.py", line 51, in connection_made
    super().connection_made(transport)
  File "/usr/lib/python3.5/asyncio/streams.py", line 238, in connection_made
    self._stream_writer)
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 660, in client_connected
    self.worker_task = asyncio_ensure_future(self.run(), loop=self.loop)
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 434, in run
    msg = yield from self.read_message()
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 456, in read_message
    frame = yield from self.read_data_frame(max_size=self.max_size)
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 511, in read_data_frame
    frame = yield from self.read_frame(max_size)
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 546, in read_frame
    self.reader.readexactly, is_masked, max_size=max_size)
  File "/usr/local/lib/python3.5/dist-packages/websockets/framing.py", line 90, in read_frame
    raise WebSocketProtocolError("Reserved bits must be 0")
websockets.exceptions.WebSocketProtocolError: Reserved bits must be 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 441, in run
    yield from self.fail_connection(1002)
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 650, in fail_connection
    yield from self.close_connection()
  File "/usr/local/lib/python3.5/dist-packages/websockets/protocol.py", line 607, in close_connection
    "in the {} state".format(self.state_name))
websockets.exceptions.InvalidState: Cannot close a WebSocket connection in the CONNECTING state
```

I've no idea what "Reserved bits must be 0" means but I'd like to stop this somehow. My best guess is that someone is trying to connect to the websocket server improperly (not using the right protocol). This could be possible since it's behind a standard load balancer and uses port 443.

Any recommendations how to throw out requests before they cause this exception? Right now we require authentication using an auth header, but I'm not sure we can check that pre-emptively before your websocket protocol handler throws this exception.

Thank you!

**Comment by aaugustin** [comment]
The really weird thing here is that websocket frames are processed while the connection is still in the CONNECTING state. This cannot happen on master anymore — see `WebSocketCommonProtocol.connection_open`.

The error message isn't great, it should state which bits have the wrong value. Then we could have a better idea of what's going on.

**Comment by jacksonofalltrades** [comment]
Hm, maybe I need to upgrade my copy of your library, then?

**Comment by aaugustin** [comment]
I intend to release version 4.0 within a few days. I hope it will make that error go away or at least make it easier to figure out what's happening there.

**Comment by jacksonofalltrades** [comment]
That's fantastic, thanks for letting me know! I will upgrade when it's ready.

**Comment by aaugustin** [comment]
FYI the new version was released, it's 4.0.1.
